core/io/logger.cpp

```diff
@@ -278,3 +278,95 @@ CompositeLogger::~CompositeLogger() {
 		memdelete(loggers[i]);
 	}
 }
+
+BufferedSingletonLogger::BufferedSingletonLogger(int p_buffer_size) {
+	buffer_size = p_buffer_size;
+}
+
+BufferedSingletonLogger::~BufferedSingletonLogger() {
+}
+
+Array BufferedSingletonLogger::flushMessageBuffer() {
+	Array return_array = message_buffer.duplicate();
+	message_buffer.clear();
+	return return_array;
+}
+
+Array BufferedSingletonLogger::flushErrorBuffer() {
+	Array return_array = error_buffer.duplicate();
+	error_buffer.clear();
+	return return_array;
+}
+
+void BufferedSingletonLogger::logv(const char *p_format, va_list p_list, bool p_err) {
+	const int static_buf_size = 512;
+	char static_buf[static_buf_size];
+	char *buf = static_buf;
+	va_list list_copy;
+	va_copy(list_copy, p_list);
+	int len = vsnprintf(buf, static_buf_size, p_format, p_list);
+	if (len >= static_buf_size) {
+		buf = (char *)Memory::alloc_static(len + 1);
+		vsnprintf(buf, len + 1, p_format, list_copy);
+	}
+	va_end(list_copy);
+	String new_message = String(buf);
+	if (message_buffer.size() >= buffer_size) {
+		message_buffer.pop_front();
+		message_buffer.pop_front();
+		message_buffer.push_back(new_message);
+		WARN_PRINT("Maximum size of buffered log has been exceeded. Discarding log messages.");
+	} else {
+		message_buffer.push_back(new_message);
+	}
+
+	if (len >= static_buf_size) {
+		Memory::free_static(buf);
+	}
+
+	if (Engine::get_singleton()->has_singleton("Engine")) {
+		Engine::get_singleton()->get_singleton_object("Engine")->emit_signal("log_message", message_buffer.size());
+	}
```

Comment:
Seems weird to ask the `Engine` singleton for the `Engine` singleton :)

Reply:
I know. But i am getting `'class Engine' has no member named 'emit_signal'` if i dont. I am not entirely sure why, but i think it is a difference between the c++ singleton and the singleton object visible from Godot.